book/src/rules/relations.md

```diff
@@ -0,0 +1,137 @@
+
+```rust
+bb0: {
+  ...          // bb0[0]
+  ...          // bb0[1]
+  goto -> bb3; // bb0[2]
+}
+
+...
+
+bb3: {
+  ... // bb3[0]
+}
+
+```
+
+we will record these input facts (as mentioned before, they'll be interned) in the `cfg_edge` relation, shown here as pseudo Rust:
+
+```rust
+cfg_edge = vec![
+  // statement at location bb0[0]:
+  (bb0-0-start, bb0-0-mid),
+  (bb0-0-mid, bb0-1-start),
+  
+  // statement at location bb0[1]:
+  (bb0-1-start, bb0-1-mid),
+  (bb0-1-mid, bb0-2-start),
+  
+  // terminator at location bb0[2]:
+  (bb0-2-start, bb0-2-mid),
+  (bb0-2-mid, bb3-0-start),
+];
+```
+
+### 2. `borrow_region`
+
+`borrow_region(origin, loan, node)`: this relation stores that the origin `origin` may refer to data from loan `loan` from the node `node` and onwards.
+
+For every borrow expression, a loan will be created and there will be a fact stored in this relation to link this loan to the origin of the borrow expression.
+
+For example
+
+```rust
+let mut a = 0;
+let r = &mut a; // this creates the loan L0
+//      ^ let's call this 'a
+```
+
+there will be a `borrow_region` fact linking `L0` to `'a` at this node. This loan will flow along the CFG and the subset relationships between origins, and the computation will require that its terms are respected or it will generate an illegal access error.
+
+### 3. `universal_region`
+
+`universal_region(origin)`: this relation stores that the origin `origin` is a "universal region"/"free region"/"placeholder region" and not defined in the MIR body we're checking.
+
+Those are the default universal origins (`'static`) and the ones defined on functions which are generic over a lifetime. For example, with
+
+```rust
+fn my_function<'a, 'b>(x: &'a u32, y: &'b u32) {
+    ...
+}
+```
+
+the `universal_region` relation will also contain facts for `'a`, and `'b`.
+
+### 4. `killed`
+
+`killed(loan, node)`: this relation stores that a prefix of the path borrowed in loan `loan` is overwritten at the node `node`.
+
+For example, with
+
+```rust
+let mut a = 1;
+let mut b = 2;
+let mut q = &mut a;
+let r = &mut *q; // loan L0 of `*q`
+// `q` can't be used here, one has to go through `r`
+q = &mut b; // killed(L0)
+// `q` and `r` can be used here
+```
+
+the loan `L0` will be `killed` by the assignment, and this fact stored in the relation. When we compute which loans origins contain along the CFG, the `killed` nodes will stop this loan's propagation to the next CFG node.
+
+### 5. `outlives`
+
+`outlives(origin1, origin2, node)`: this relation stores that the origin `origin1` outlives origin `origin2` at the node `node`.
+
+This is the standard Rust syntax `'a: 'b` where the *lifetime* `'a` outlives the lifetime `'b`. From the node of view of origins as sets of loans, this is seen as a subset-relation: with all the loans in `'a` flowing into `'b`, `'a` contains a subset of the loans `'b` contains. 
```

Comment:
"from the node of view" is an overzealous s/point/node/g — which I'll fix when switching back to "points"

Reply:
typo fixed